cpp/include/raft/sparse/matrix/preprocessing.cuh

```diff
@@ -0,0 +1,568 @@
+ * @param[in] columns
+ *   array coo representation of columns for non zero values.
+ * @param[in] values
+ *   array coo representation of non zero values.
+ * @param[in] nnz
+ *   The number of non-zero values in the coo array representations.
+ * @param[in] results
+ *   array that is the size of the csr_in nnz that will store the encoded
+ *      transform values.
+ * @param[in] bm25_on
+ *   When true calculates bm25 encoding values, otherwise tfidf.
+ * @param[in] k_param
+ *   bm25 okapi optimization parameter k1
+ * @param[in] b_param
+ *   bm25 okapi optimization parameter b
+ * */
+template <typename ValueType, typename IndexType>
+void SparseEncoder<ValueType, IndexType>::transform(
+  raft::resources const& handle,
+  raft::device_vector_view<IndexType, int64_t> rows,
+  raft::device_vector_view<IndexType, int64_t> columns,
+  raft::device_vector_view<ValueType, int64_t> values,
+  IndexType nnz,
+  ValueType* results,
+  bool bm25_on,
+  float k_param,
+  float b_param)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+
+  int maxLimit   = nnz;
+  int blockSize  = (maxLimit < 128) ? maxLimit : 128;
+  int num_blocks = (maxLimit + blockSize - 1) / blockSize;
+  float avgIdLen = (ValueType)fullIdLen / numRows;
+  int* counts;
+  cudaMallocManaged(&counts, maxLimit * sizeof(IndexType));
+  cudaMemset(counts, 0, maxLimit * sizeof(IndexType));
+  raft::sparse::op::coo_sort(IndexType(rows.size()),
+                             IndexType(columns.size()),
+                             IndexType(values.size()),
+                             rows.data_handle(),
+                             columns.data_handle(),
+                             values.data_handle(),
+                             stream);
+  raft::sparse::matrix::detail::_scan<<<num_blocks, blockSize>>>(rows.data_handle(), nnz, counts);
+  raft::sparse::matrix::detail::_transform<<<num_blocks, blockSize>>>(rows.data_handle(),
+                                                                      columns.data_handle(),
+                                                                      values.data_handle(),
+                                                                      featIdCount,
+                                                                      counts,
+                                                                      results,
+                                                                      numRows,
+                                                                      avgIdLen,
+                                                                      k_param,
+                                                                      b_param,
+                                                                      nnz,
+                                                                      numFeats,
+                                                                      bm25_on);
+  cudaFree(counts);
```

Comment:
A concern is that we should take care to use these explicitly and implicitly device-sync APIs, which will break the parallel pipeline in the end-user environment and hurt performance.